zulip_bots/zulip_bots/zulip_bot_output.py

```diff
@@ -0,0 +1,124 @@
+        error_message = """
+You must either specify the name of an existing bot or
+specify a path to the file (--path-to-bot) that contains
+the bot handler class.
+If you provide both: the name and the path to the bot,
+make sure that both point to the same bot.
+"""
+        parser.error(error_message)
+
+    return options
+
+def main():
+    # type: () -> None
+    options = parse_args()
+    bot_name = options.name
+    if options.path_to_bot:
+        if options.provision:
+            bot_dir = os.path.dirname(os.path.abspath(options.path_to_bot))
+            provision_bot(bot_dir, options.force)
+        lib_module = import_module_from_source(options.path_to_bot, name=bot_name)
+    elif options.name:
+        if options.provision:
+            current_dir = os.path.dirname(os.path.abspath(__file__))
+            bots_parent_dir = os.path.join(current_dir, "bots")
+            bot_dir = os.path.join(bots_parent_dir, options.name)
+            provision_bot(bot_dir, options.force)
+        lib_module = import_module('zulip_bots.bots.{bot}.{bot}'.format(bot=bot_name))
+
+    message = {'content': options.message, 'sender_email': 'foo_sender@zulip.com'}
+    message_handler = lib_module.handler_class()
+
+    with patch('zulip_bots.lib.ExternalBotHandler') as mock_bot_handler:
+        def get_config_info(bot_name, section=None, optional=False):
+            # type: (str, Optional[str], Optional[bool]) -> Dict[str, Any]
+            conf_file_path = os.path.realpath(os.path.join(
+                'zulip_bots', 'bots', bot_name, bot_name + '.conf'))
+            section = section or bot_name
+            config = configparser.ConfigParser()
+            try:
+                with open(conf_file_path) as conf:
+                    config.readfp(conf)  # type: ignore
+            except IOError:
+                if optional:
+                    return dict()
+                raise
+            return dict(config.items(section))
+
+        mock_bot_handler.get_config_info = get_config_info
+        if hasattr(message_handler, 'initialize') and callable(message_handler.initialize):
+            message_handler.initialize(mock_bot_handler)
+
+        mock_bot_handler.send_reply = MagicMock()
+        mock_bot_handler.send_message = MagicMock()
+        message_handler.handle_message(
+            message=message,
+            bot_handler=mock_bot_handler,
+            state_handler=StateHandler()
+        )
+        print("On sending ", options.name, " bot the following message:\n\"", options.message, "\"")
+        if mock_bot_handler.send_reply.called:
+            print("\nThe bot gives the following output message:\n\"", list(mock_bot_handler.send_reply.call_args)[0][1], "\"")
+        elif mock_bot_handler.send_message.called:
+            print("\nThe bot sends the following output to zulip:\n\"", list(mock_bot_handler.send_message.call_args)[0][0], "\"")
```

Comment:
I guess you can just assign one message string to some `message` variable and then do 
```
print("\nThe bot sends the following message:\n\"", list(mock_bot_handler.send_reply.call_args)[0][1], "\"")
```

Reply:
As a matter of fact, I have purposefully kept the 2 strings different. Although very slight difference.
`
if mock_bot_handler.send_reply.called:
            print("\nThe bot gives the following output message:\n\"", list(mock_bot_handler.send_reply.call_args)[0][1], "\"")
`
Just outputs the message content.

However,
`
elif mock_bot_handler.send_message.called:
            print("\nThe bot sends the following output to zulip:\n\"", list(mock_bot_handler.send_message.call_args)[0][0], "\"")
` 
prints the entire message with the stream name (for followup bot)

Reply:
I focussed all my thoughts on the commit message: https://github.com/zulip/python-zulip-api/pull/90/commits/d447d33e79c09fffb1c8522579f75ee3245140a8 .

I'll add a comment for the same in the file. Thanks for pointing it out. :)

Reply:
Hmm, I see. What's the advantage of handling `send_reply` and `send_message` differently in this way?

Reply:
When `send_reply` is called from any bot's code, it simply calls it by passing (message, response).
Here message is the origin message sent by the user and `response` is the bot's reply after processing message['content']. The bot does not interfere on deciding if this bot response will be private or in a stream, or which topic. The reply is simply sent through the channel we received 'message'.

So, if we want to test bot output, simply the message will suffice in case `send_reply` is called (as the bot does not interfere with any other message fields apart from content.


This is not the case with `send_message`. If `send_message` is directly called by any bot's code, then it can edit/modify all the fields of the response message, hence we need to output all the fields to check its output.
This can simply be understood by taking an example of `follow_up` bot. It uses `send_message` to send message to a different stream with 'sender's name' as the topic.

I hope I have put forward my thoughts clearly. Makes sense?

Reply:
Makes sense, thanks for the detailed explanation! I guess  I'll understand all of this better once I've actually run the command successfully :)

Reply:
So, if we output all the fields of the reply message for the bot's that use 'send_reply', that would simply increase our effort of inputting them as well. These do not play any role in the bot logic, so it will basically be useless in our testing aka getting the bot output.

However this does play in bots making use of `send_message`, as while testing if the bot is editing/modifying some response fields, we'll want to know that.

Does it make sense? :)